src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+﻿﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+namespace Microsoft.Quantum.QsCompiler
+{
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+    // Type Parameters are frequently referenced by the callable of the type parameter followed by the name of the specific type parameter.
+    using TypeParameterName = Tuple<QsQualifiedName, NonNullable<string>>;
+    using TypeParameterResolutions = ImmutableDictionary</*TypeParameterName*/ Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+
+    /// <summary>
+    /// Summarizes a series of type parameter resolution dictionaries, ResolutionsInSummary, into one
+    /// resolution dictionary, CombinedTypeParameterResolutions, containing the ultimate type resolutions
+    /// for all the type parameters found in the dictionaries. Validation is done on the resolutions,
+    /// which can be checked through the IsValidSummary, SummarizesOverConflictingResolution, and
+    /// SummarizesOverParameterConstriction flags.
+    /// </summary>
+    public class TypeResolutionSummary
```

Comment:
How will this class be used? I'm not sure I understand the design and purpose of the properties it exposes without knowing what they are needed for.

Reply:
I guess I won't use the conceptualization of it being a summary since that is not the direction we are going in anymore.
It takes a series of type parameter resolutions and combines them into one resolution mapping. While doing this, it sets some validation flags so the user can tell if something went wrong during the combining, though they may still want the combined resolution depending on the situation. Since getting the right resolutions for a TypedExpression can be tedious, the public constructor gets all the resolutions for you from the provided TypedExpression. This can be useful, for example, if you want to know what are all the type parameter resolutions for a given call-like expression.

Reply:
Is it something we expect users to interact with using an API or is it only for the call graph? What will users or the call graph do with information like `IsValid`, `CombinesOverConflictingResolution`, and `CombinesOverParameterConstriction`?

Reply:
I guess specifically I'm wondering:

* What is the difference between `IsValid` and the diagnostics generated by the compiler for invalid type parameters?
* It seems like `CombinesOverConflictingResolution` and `CombinesOverParameterConstriction` both imply `not IsValid`. When is it needed to know specifically that there was a conflicting resolution or a constricted type, versus looking at `IsValid` or the compiler diagnostics?

Reply:
This is something that coders using the QDK can use to on TypedExpressions to get the complete picture of type parameter resolutions in the expression. When writing transformations for the QDK, users may need to create custom TypedExpressions from scratch or from other components acquired from the syntax tree. The validation flags can be used to check if the created TypedExpression has correct type parameter resolutions.

Reply:
I missed your last message. To your first question, I think I've answered that to the best of my ability. Since it is a public class, I can't account for you here how any user of the QDK will use it. For your second question: Whatever diagnostics you are referring to must not happen on the level that this class deals with as they are not being raised when testing invalid scenarios. It is possible that the diagnostics are specific to individual resolution dictionaries, while this class checks for combinations between dictionaries. For your third question: the `IsValid` is derived from the combination of the other two flags. While `IsValid` will probably be used more frequently, when someone does have an invalid series of type parameters, or TypedExpression, they will likely want to know more about what is making it invalid. They can being this by checking the other two flags for what quality specifically was violated. And they will be able to look at the input resolution dictionaries with the `IndependentResolutionDictionaries` since they will have just started with a TypedExpression.

Reply:
To design a public API, to me it seems like it is important to have concrete use cases in mind. For example, are there any specific examples of transformations that a user would want to write that would be difficult to do without this tool? Have these examples come up in practice - is there a feature request for this? Is there a use case for this within the compiler or runtime, and if so can it be designed as an internal tool instead of public?

Sorry for all of the questions, but the reason I'm asking is because it seems like there are multiple ways to design this, and I think that the design should be informed by those use cases in order to best accomplish the task. Since I'm not sure what those use cases are, I'm not able to review the design in the PR.

Reply:
Let's take this offline.

Reply:
To summarize our conversation: I think there is a lot of context here that I'm not familiar with, and I'm not sure how this PR fits in with the big picture in the compiler. So I'll decline to review the PR since I am not confident saying one way or the other what the design should be.